src/Pretzel/Program.cs

```diff
@@ -1,89 +1,160 @@
-using Pretzel.Commands;
-using Pretzel.Logic.Commands;
-using Pretzel.Logic.Extensions;
 using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Invocation;
 using System.Composition;
 using System.Composition.Hosting;
 using System.IO;
 using System.IO.Abstractions;
 using System.Linq;
 using System.Reflection;
+using System.Threading.Tasks;
+using Pretzel.Commands;
+using Pretzel.Logic.Extensions;
 
 namespace Pretzel
 {
+    [Export]
     internal class Program
     {
-        [Import]
-        public CommandCollection Commands { get; set; }
+#pragma warning disable S2223 // Non-constant static fields should not be visible
+        internal static IFileSystem fileSystem = new FileSystem();
 
-        [Export]
-        public IFileSystem FileSystem { get; set; } = new FileSystem();
+        internal static Option[] GlobalOptions = new[]
+        {
+            new Option("--debug", "Enable debugging")
+            {
+                Argument = new Argument<bool>()
+            },
+            new Option("--safe", "Disable custom plugins")
+            {
+                Argument = new Argument<bool>()
+            },
+            new Option(new[] { "--source", "-source", "-s", "--s" }, "The path to the source site (default current directory)")
+            {
+                Argument = new Argument<string>(() => fileSystem.Directory.GetCurrentDirectory())
+            }
+        };
+#pragma warning restore S2223 // Non-constant static fields should not be visible
 
-        [Export("SourcePath")]
-        public string SourcePath { get; }
+        [Import]
+        public CommandCollection CommandCollection { get; set; }
 
-        private BaseParameters parameters { get; }
+        [Export]
+        public IFileSystem FileSystem { get; set; } = fileSystem;
 
-        public Program()
+        private static async Task<int> Main(string[] args)
         {
-            parameters = BaseParameters.Parse(Args, FileSystem);
-            SourcePath = parameters.Path;
-        }
+            var rootCommand = new RootCommand
+            {
+                TreatUnmatchedTokensAsErrors = false
+            };
 
-        private static string[] Args;
-        private static void Main(string[] args)
-        {
-            Args = args;
-            var program = new Program();
+            args = PatchSourcePath(args);
 
-            program.InitializeTrace();
-            Tracing.Info("starting pretzel...");
-            Tracing.Debug("V{0}", Assembly.GetExecutingAssembly().GetName().Version);
+            foreach (var option in GlobalOptions)
```

Comment:
I did this to support the built in help command even if a command failes. Cause it's used in the startup phase (we need the source path for plugins, and debug and safe to control it's behavior). They map to the arugments in the root command handler `CommandHandler.Create(async (bool debug, bool safe, string source)`. [Program](https://github.com/Code52/pretzel/pull/347/files#diff-40e719901a3e1c2feb8c537746f27bfcR59)

Reply:
Since they are added to the subcommands, I don't think that they are needed in the RootCommand for the help.
And they can be used directly through their `Parse` method without the need to attach them to a command.

Reply:
I'm not sure about this:

```cs
   var parseResult = new Parser(GlobalOptions).Parse(args);

            if (parseResult.Errors.Count > 0)
            {
                foreach (var error in parseResult.Errors)
                {
                    // Can' use Trace here cause we need Debug?!
                }
                return -1;
            }

            rootCommand.Handler = CommandHandler.Create(async (bool debug, bool safe, string source) =>
            {
                try
                {
                    InitializeTrace(debug);
                    Tracing.Info("starting pretzel...");
```

Reply:
I could initialize Trace with debug first, and reinitialize later with the argument provided by the user. 🤷‍♂️

Reply:
Since debug only sets the minimal trace level to debug I don't think it is needed to trace the parse errors.
`InitializeTrace` can be separated into two parts: 
1. `Tracing.SetTrace(ConsoleTrace.Write);` as soon as possible
2. `Tracing.SetMinimalLevel(TraceLevel.Debug);` once the parsing is done

Reply:
Good to me!

Reply:
Better?

Reply:
Excellent :)